applications/FreeSurfaceApplication/custom_utilities/edgebased_levelset_substep.h

```diff
@@ -51,2028 +51,2012 @@
-                        if ( (static_cast<int> (j_neighbour) > i_node) && (flag == 0) )
+                        // loop over all nodes
+                        // 	    for (unsigned int i_node = 0; i_node < n_nodes; i_node++) {
+                        // flag for considering diagonal matrix elements
+                        bool flag = 0;
+                        // loop over all neighbours
+                        for (unsigned int csr_index = mr_matrix_container.GetRowStartIndex()[i_node]; csr_index != mr_matrix_container.GetRowStartIndex()[i_node + 1]; csr_index++)
                         {
-                            //add diagonal/nodal contribution
-                            mL.push_back (i_node, i_node, 0.0);
-                            flag = 1;
+                            // get global index of neighbouring node j
+                            unsigned int j_neighbour = mr_matrix_container.GetColumnIndex()[csr_index];
+                            // define matrix structure row by row (the order does matter!)
+                            if ((static_cast<int>(j_neighbour) > i_node) && (flag == 0))
+                            {
+                                // add diagonal/nodal contribution
+                                mL.push_back(i_node, i_node, 0.0);
+                                flag = 1;
+                            }
+                            // add non-diagonal/edge contribution
+                            mL.push_back(i_node, j_neighbour, 0.0);
                         }
-                        //add non-diagonal/edge contribution
-                        mL.push_back (i_node, j_neighbour, 0.0);
+                        // if diagonal element is the last non-zero element of the row
+                        if (flag == 0)
+                            mL.push_back(i_node, i_node, 0.0);
                     }
-                    //if diagonal element is the last non-zero element of the row
-                    if (flag == 0)
-                        mL.push_back (i_node, i_node, 0.0);
                 }
             }
+            // compute minimum length of the surrounding edges
+            CalculateEdgeLengths(mr_model_part.Nodes());
+            // set the pressure projection to the body force value
+            array_1d<double, 3> temp = mRho * mBodyForce;
+            for (ModelPart::NodesContainerType::iterator inode = mr_model_part.NodesBegin();
+                 inode != mr_model_part.NodesEnd();
+                 inode++)
+                inode->FastGetSolutionStepValue(PRESS_PROJ) = temp;
+            mr_matrix_container.FillScalarFromDatabase(POROSITY, mEps, mr_model_part.Nodes());
+            // verify that neither h_min nor havg are 0
+            for (unsigned int i_node = 0; i_node < mHmin.size(); i_node++)
+            {
+                if (mHmin[i_node] < 1e-20)
+                    KRATOS_THROW_ERROR(std::logic_error, "hmin too small on node ", i_node + 1)
```

Comment:
```suggestion
            {
                KRATOS_ERROR_IF(mHmin[i_node] < 1e-20) << "hmin too small on node " << i_node + 1 << std::endl;
```
You can use the `KRATOS_ERROR_IF` macro for these. Besides the tolerance is too small, I'd use something close but not smaller than the machine precision (e.g. 1.0e-15).

Reply:
Done